### DISCUSSION TITLE
Custom formatting for really long "special" String?

### DISCUSSION THREAD

**Discussion by nielsbasjes** [charleskorn/kaml#594] [Q&A]
Hi,

I have a usecase where I have a list of registers values that I want to read and write to a yaml file.

To illustrate a snippet of my input (manually created to look like this)
```
  registers: |
      436A A2CB 436A 5E5E 436B CFF8 4060 354B 4048 A3A4 4081 3697 4405 C7E8 4409 8DA7
      4437 E384 444C A922 4437 4283 446D 7231 441A E1FC 43F2 2F0C 4416 37EA 3F27 5706
      3F40 26C9 3F46 41FA 4244 BD0D 4225 720D 421C FE2D 436A F060 0000 0000 4063 C209
      412A D187 0000 0000 44E3 9C8A 0000 0000 451B E5F6 0000 0000 44D5 18B6 3F3A E146
      0000 0000 422C 7797 0000 0000 4247 99A6 4109 BA5E 0000 0000 40B1 BA5E 0000 0000
```
That I can read with this
```
@Serializable
data class SchemaTestRegisters(
...
    /** The register values used for this test. */
    val registers: String,
)
```
After that a parser will convert this String into something my code will work with.

I would like the reverse to be "the same" in terms of readability of the produced output. 

When I simply write the data back to yaml (using a simple toString that creates a single line) I get this (snippet of full output):
```
    registers: "436A A2CB 436A 5E5E 436B CFF8 4060 354B 4048 A3A4 4081 3697 4405 C7E8\
      \ 4409 8DA7 4437 E384 444C A922 4437 4283 446D 7231 441A E1FC 43F2 2F0C 4416\
      \ 37EA 3F27 5706 3F40 26C9 3F46 41FA 4244 BD0D 4225 720D 421C FE2D 436A F060\
      \ 0000 0000 4063 C209 412A D187 0000 0000 44E3 9C8A 0000 0000 451B E5F6 0000\
...
      \ 0000 4247 C61B 4109 BA5E 0000 0000 40B1 BA5E 0000 0000"
```

At the technical level this is correct and works fine.

Yet I would like to have it formatted as shown in the original input; a block with a '|' , no quotes and on every 16th element a newline (because of hex addresses).
I can provide the desired values (now a too simple toString) in any format to meet this.

How do I massage my code to produce this output?

**Comment by nielsbasjes** [comment]
I have tried to insert newlines and that did not have the desired effect:
```
    registers: "\n436A A2CB 436A 5E5E 436B CFF8 4060 354B 4048 A3A4 4081 3697 4405\
      \ C7E8 4409\n8DA7 4437 E384 444C A922 4437 4283 446D 7231 441A E1FC 43F2 2F0C\
      \ 4416 37EA 3F27\n5706 3F40 26C9 3F46 41FA 4244 BD0D 4225 720D 421C FE2D 436A\
      \ F060 0000 0000 4063\nC209 412A D187 0000 0000 44E3 9C8A 0000 0000 451B E5F6\
...
```

**Comment by charleskorn** [comment]
Formatting the output in the way you suggest is currently not supported by Kaml. 

It might be possible to add an annotation that tells Kaml to use a particular style when writing the value - would you be interested in putting together a PR to do this?

**Reply by nielsbasjes** [reply]
I'll have a look if I can come up with a clean way of doing that.

**Reply by nielsbasjes** [reply]
As I'm very new to the internals of Kotlin serialization: Would this be possible with an existing annotation or would this need a Kaml specific annotation? Do you have something specific in mind?

**Reply by charleskorn** [reply]
It would need to be a new annotation, specific to Kaml. You might be able to look at the existing `YamlComment` annotation for inspiration.

**Reply by nielsbasjes** [reply]
I love to get your comments on this first attempt https://github.com/charleskorn/kaml/pull/596
